L1Trigger/TrackerTFP/interface/DataFormats.h

```diff
@@ -0,0 +1,672 @@
+  // track trigger process order
+  constexpr std::initializer_list<Process> Processes = {Process::fe, Process::dtc, Process::pp, Process::gp, Process::ht, Process::mht, Process::sf, Process::kfin, Process::kf, Process::dr};
+  // conversion: Process to int
+  inline constexpr int operator+(Process p) { return static_cast<int>(p); }
+  // conversion: Variable to int
+  inline constexpr int operator+(Variable v) { return static_cast<int>(v); }
+  // increment of Process
+  inline constexpr Process operator++(Process p) { return Process(+p + 1); }
+  // increment of Variable
+  inline constexpr Variable operator++(Variable v) { return Variable(+v + 1); }
+
+  //Base class representing format of a variable
+  class DataFormat {
+  public:
+    DataFormat(bool twos) : twos_(twos), width_(0), base_(1.), range_(0.) {}
+    ~DataFormat() {}
+    // converts int to bitvector
+    TTBV ttBV(int i) const { return TTBV(i, width_, twos_); }
+    // converts double to bitvector
+    TTBV ttBV(double d) const { return TTBV(d, base_, width_, twos_); }
+    // extracts int from bitvector, removing these bits from bitvector
+    void extract(TTBV& in, int& out) const { out = in.extract(width_, twos_); }
+    // extracts double from bitvector, removing these bits from bitvector
+    void extract(TTBV& in, double& out) const { out = in.extract(base_, width_, twos_); }
+    // extracts double from bitvector, removing these bits from bitvector
+    void extract(TTBV& in, TTBV& out) const { out = in.slice(width_, twos_); }
+    // attaches integer to bitvector
+    void attach(const int i, TTBV& ttBV) const { ttBV += TTBV(i, width_, twos_); }
+    // attaches double to bitvector
+    void attach(const double d, TTBV& ttBV) const { ttBV += TTBV(d, base_, width_, twos_); }
+    // attaches bitvector to bitvector
+    void attach(const TTBV bv, TTBV& ttBV) const { ttBV += bv; }
+    // converts int to double
+    double floating(int i) const { return (i + .5) * base_; }
+    // converts double to int
+    int integer(double d) const { return std::floor(d / base_); }
+    // converts double to int and back to double
+    double digi(double d) const { return floating(integer(d)); }
+    // converts binary integer value to twos complement integer value
+    int toSigned(int i) const { return i - std::pow(2, width_) / 2; }
+    // converts twos complement integer value to binary integer value
+    int toUnsigned(int i) const { return i + std::pow(2, width_) / 2; }
+    // returns false if data format would oferflow for this double value
+    bool inRange(double d) const { return d >= -range_ / 2. && d < range_ / 2.; }
+    // returns false if data format would oferflow for this int value
+    bool inRange(int i) const { return inRange(floating(i)); }
+    // true if twos'complement or false if binary representation is chosen
+    bool twos() const { return twos_; }
+    // number of used bits
+    int width() const { return width_; }
```

Comment:
Suggest name nbits() rather than width(), since latter could be misinterpreted as "width of the region being digitized, i.e. range()".

Reply:
I try to keep s/w similar to f/w. I will consider changing that in future but not in the next months.